### ISSUE TITLE
hashed switch commitments (thread resurrection)

### ISSUE THREAD

**Issue by antiochp** [mimblewimble/grin#590]
Concerns raised that current "switch commit" implementation does not fully give us the desired properties.

We likely need to discuss and agree on an approach here that allows us to - 
  * have switch commits that do what we desire them to do
  * still allow us to lock coinbase outputs (or find an alternative way to do this)
  * still allow us "easy" wallet restore from seed functionality

https://lists.launchpad.net/mimblewimble/msg00373.html
(msg from Tim Ruffing)

I need to resurrect this thread.

What's currently implemented in Grin [1] is 
  (vG + bH), BLAKE2(bJ), 
where the hash is truncated to 160 bits.

Hashed switch commitments are supposed to give you two main properties:
   1. Switch binding: even if you have unlimited computation power from
      tomorrow on, commitments created today are still binding if fully
      verified. This is the original property from the switch commitments
      paper [2] and it means that you can still spend your money after the
      switch to full verification (but you may lose privacy).
   2. Privacy for "old" outputs, which have been spent long ago, is
      retained even against quantum computers.
   3. Opt-out money, opt-in privacy: If your privacy is more important
      than your money, then you can decide not to reveal the pre-image of
      the hash. Then you cannot spend the money but you retain privacy
      even against post-quantum attackers.                  (This was my
      understanding, let's call it "3a". Andrew and maybe the others
      involved in the conversation had a different variant "3b" in mind:
      If you prefer privacy, then you commit to ((vG + bH), r) in the
      first place, where r is just random. This requires users to decide
      upfront if they want money or privacy but "They will have plenty of
      warning before any such fork, so  they can either move their coins
      away or move them to upgradable outputs, taking the privacy hit
      voluntarily and knowingly and only for their most recent history."

It turns out that the implemented variant just fulfills 1 and 3b [3].
Property 2 is trivially broken and it kind of scares me that I had
overlooked this:
Given the commitment ((vG + bH), BLAKE2(bJ)), guess v, compute b (by
solving dlog), and check if the second element is indeed BLAKE2(bJ)).

If we want all properties to hold, then the following construction
works:
  (vG + bH), BLAKE2(bJ, r),
where r is 256 bits and the hash is 256 bits as well.

Then 1 holds only computationally against quantum computers [4]. Even
in a quantum world, you need 2^128 operations to find a second preimage
in the hash. (Note that you have a quantum computer only after the
switch, you so cannot find a collision upfront.)

2 and 3a hold for the same reason: If you don't reveal r, then even a
quantum computer needs about 2^128 operations to compute it, and
without r you learn nothing about x [5].

So I propose that we switch to this new variant, pun intended.

Best,
Tim

**Comment by antiochp** [comment]
Followup regarding `lock_heights` and `wallet restore` - 

Just to add a couple of bits of info to this discussion -

We were actually experimenting with including a secret key as part of the blake2 hashing - using the lock_height of coinbase outputs in the switch commit hash here (not merged, still under discussion, parked due to testnet1) -
https://github.com/mimblewimble/grin/pull/215

Specifically -
https://github.com/antiochp/grin/blob/439dca94ec3970b7821b9fa78487521db1e922f1/core/src/core/transaction.rs#L545-L546

> proposed switch commit hash: blake2(h, rJ) where h the lock_height is used as the secret key to the hash function.

This was an attempt at ensuring the associated lock_height could not be altered once the coinbase output was on the block chain.

Also - I believe we are currently using the switch commit hash as a convenient way of identifying "our" outputs in the UTXO set for wallet recovery.
This is trivial currently as we know rJ from our wallet private keys and we can quickly check if the switch commit hash was generated by us.

None of this is to say I disagree - just wanted to make sure these were both visible as well.

**Comment by antiochp** [comment]
https://lists.launchpad.net/mimblewimble/msg00377.html
(Followup from Tim Ruffing)

Hm, if I understand this correctly, then I think this throws together
two unrelated concepts. With this proposal, users are forced to open
their commitments (and give up some privacy) just to realize some
unrelated functionality, i.e., timelocks.

If we want to store that in the same hash to save space, which is
probably a good idea, then I think the way to go is something like
  hash(data||hash(bJ||r))
where data can be anything, e.g., lock_height.

Then, if you show bJ, you are forced to show data as well but this is
fine:
If you show bJ, then you want to spend, and then you need to show the
lock_height anyway.

Regarding identifying outputs:
r can be derived from the wallet key as well, so that's not an issue, I
think.

Best.
Tim

**Comment by antiochp** [comment]
We have decided __not__ to use the switch commit hash for encoding lock_heights.
See #615 for the impl details.

So we no longer need to account for this with the proposed switch commit impl above.

We _do_ still want to account for `wallet restore` being able to identify our own outputs based on the switch commit hash.

**Comment by antiochp** [comment]
Closing - resolved (at least as discussed) in #648.
